pkg/template/controller/templateinstance_controller.go

```diff
@@ -1,103 +1,202 @@
 
-func (c *TemplateInstanceController) Run(stop <-chan struct{}) {
-	c.controller.Run(stop)
+func (c *TemplateInstanceController) getTemplateInstanceCopy(key string) (*templateapi.TemplateInstance, error) {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return nil, err
+	}
+
+	templateInstance, err := c.lister.TemplateInstances(namespace).Get(name)
+	if err != nil {
+		return nil, err
+	}
+
+	templateInstanceCopy, err := kapi.Scheme.DeepCopy(templateInstance)
+	if err != nil {
+		return nil, err
+	}
+
+	return templateInstanceCopy.(*templateapi.TemplateInstance), nil
 }
 
-func (c *TemplateInstanceController) handle(templateInstance *templateapi.TemplateInstance) error {
+func (c *TemplateInstanceController) sync(key string) error {
+	templateInstance, err := c.getTemplateInstanceCopy(key)
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
 	for _, condition := range templateInstance.Status.Conditions {
 		if condition.Type == templateapi.TemplateInstanceReady && condition.Status == kapi.ConditionTrue ||
 			condition.Type == templateapi.TemplateInstanceInstantiateFailure && condition.Status == kapi.ConditionTrue {
 			return nil
 		}
 	}
 
-	err := c.provision(templateInstance)
-	if err == nil {
-		templateInstance.Status.Conditions = []templateapi.TemplateInstanceCondition{
-			{
+	provisionErr := c.provision(templateInstance)
+
+	for try := 0; try < updateRetries; try++ {
+		templateInstance.Status.Conditions = templateapi.FilterTemplateInstanceCondition(templateInstance.Status.Conditions, templateapi.TemplateInstanceReady)
+		templateInstance.Status.Conditions = templateapi.FilterTemplateInstanceCondition(templateInstance.Status.Conditions, templateapi.TemplateInstanceInstantiateFailure)
+
+		if provisionErr == nil {
+			templateInstance.Status.Conditions = append(templateInstance.Status.Conditions, templateapi.TemplateInstanceCondition{
 				Type:               templateapi.TemplateInstanceReady,
 				Status:             kapi.ConditionTrue,
 				LastTransitionTime: metav1.Now(),
 				Reason:             "Created",
-			},
-		}
+			})
 
-	} else {
-		templateInstance.Status.Conditions = []templateapi.TemplateInstanceCondition{
-			{
+		} else {
+			templateInstance.Status.Conditions = append(templateInstance.Status.Conditions, templateapi.TemplateInstanceCondition{
 				Type:               templateapi.TemplateInstanceInstantiateFailure,
 				Status:             kapi.ConditionTrue,
 				LastTransitionTime: metav1.Now(),
 				Reason:             "Failed",
-				Message:            err.Error(),
-			},
+				Message:            provisionErr.Error(),
+			})
+		}
+
+		_, err = c.templateclient.TemplateInstances(templateInstance.Namespace).UpdateStatus(templateInstance)
+		if err == nil {
+			break
+		}
+
+		utilruntime.HandleError(fmt.Errorf("TemplateInstance status update failed (%d/%d): %v", try+1, updateRetries, err))
+
+		if apierrors.IsConflict(err) && try+1 < updateRetries {
```

Comment:
@jim-minter so the big issue here is that if we retry the event, we'll get create failures because the objects already exist, right?  But we should be able to tell that we created those objects based on their ownerrefs and ignore the create failures and proceed to the status update.

So i'm inclined to say we can just retry the event as @deads2k suggests (if we add the other logic to be more intelligent about handling create failures), unless i'm forgetting something?